jax/_src/scipy/signal.py

```diff
@@ -146,3 +151,336 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=None):
+  left_end = axis_slice(x, start=0, stop=1, axis=axis)
+  left_ext = axis_slice(x, start=n, stop=0, step=-1, axis=axis)
+  right_end = axis_slice(x, start=-1, axis=axis)
+  right_ext = axis_slice(x, start=-2, stop=-(n + 2), step=-1, axis=axis)
+  ext = jnp.concatenate((2 * left_end - left_ext,
+                         x,
+                         2 * right_end - right_ext),
+                         axis=axis)
+  return ext
+
+
+_spectral_helper_mode = ('psd', 'stft')
+
+
+def _spectral_helper(x, y,
+                     fs=1.0, window='hann', nperseg=None, noverlap=None,
+                     nfft=None, detrend_type='constant', return_onesided=True,
+                     scaling='density', axis=-1, mode='psd', boundary=None,
+                     padded=False):
+  """LAX-backend implementation of `scipy.signal._spectral_helper`.
+
+  Unlike the original helper function, `y` can be None for explicitly
+  indicating auto-spectral (non cross-spectral) computation.  In addition to
+  this, `detrend` argument is renamed to `detrend_type` for avoiding internal
+  name overlap.
+  """
+  if mode not in _spectral_helper_mode:
+    raise ValueError(f"Unknown value for mode {mode}, "
+                     f"must be one of: {_spectral_helper_mode}")
+
+  def make_pad(mode, **kwargs):
+    def pad(x, n, axis=-1):
+      pad_width = [(0, 0) for unused_n in range(x.ndim)]
+      pad_width[axis] = (n, n)
+      return jnp.pad(x, pad_width, mode, **kwargs)
+    return pad
+
+  boundary_funcs = {
+      'even': make_pad('reflect'),
+      'odd': odd_ext,
+      'constant': make_pad('edge'),
+      'zeros': make_pad('constant', constant_values=0.0),
+      None: lambda x, *args, **kwargs: x
+  }
+
+  if boundary not in boundary_funcs:
+    raise ValueError(
+        f"Unknown boundary option '{boundary}', "
+        f"must be one of: {list(boundary_funcs.keys())}")
+
+  # Check/ normalize inputs
+  nperseg = jax.core.concrete_or_error(int, nperseg,
+                                       "nperseg of windowed-FFT")
+  noverlap = jax.core.concrete_or_error(int, noverlap,
+                                        "noverlap of windowed-FFT")
+  if nfft is not None:
+    nfft = jax.core.concrete_or_error(int, nfft,
+                                      "nfft of windowed-FFT")
+  axis = jax.core.concrete_or_error(int, axis,
+                                    "axis of windowed-FFT")
+
+  _check_arraylike("_spectral_helper", x)
+  if x is y:
+    y = None  # For SciPy compatibility (but doesn't work in general with jit)
+  if y is not None:
+    _check_arraylike("_spectral_helper", y)
+  axis = operator.index(axis)
+  x = jnp.asarray(x)
+  if y is not None:
+    y = jnp.asarray(y)
+    outdtype = jnp.result_type(x, y, np.complex64)
+  else:
+    outdtype = jnp.result_type(x, np.complex64)
+
+  if y is not None and mode != 'psd':
+    raise ValueError("two-argument mode is available only when mode=='psd'")
```

Comment:
In CSD case, mode == 'psd'. Therefore, ValueError will not be raised.  On the other hand, if we call jit(csd)(x, x), we don't get a computation short-cut as we have in "csd(x, x)", but at least the results should be identical.
Might be better if I rephrase the condition statement to be `mode == "stft"`

Reply:
Okay, I think now I understand it's more complicated than I originally thought.
The output dtype of `_spectral_helper` with `mode=='psd'  is different depending on whether `x is y` or not, because of L423 below:
https://github.com/google/jax/commit/43462d8c934bd63ade73aad81167071f38d09d4d#diff-70fbf1556869ea4771d5a99c679f20051a98d0d1dc15629440e95813b0abc468R423
and this issue breaks transparency of JIT.

In fact, the original scipy function has the similar type issue:
```
>>> sp.signal.csd(x, x)[1].dtype
dtype('float64')
>>> sp.signal.csd(x, x.copy())[1].dtype
dtype('complex128')
```
and JIT makes it visible even when `csd(x, y)` is called.

Reply:
I think computing CSD with the same variable is anyway an artificial corner case.  So, we can safely assume that `csd` returns complex array.  I'll document this behavior in the docstring.

Reply:
Added document, and users can now give `None` as the second parameter of `csd` for explicitly instructing to follow the behavior when `x is y`.